File: leetcode/java/0104. Maximum Depth of Binary Tree.java
package leetcode.java;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time complexity: O(N), where N is the number of nodes
// Space complexity: O(N)

class Solution {
    public int maxDepth(TreeNode root) {
        return this.dfs(root);
    }

    public int dfs(TreeNode source) {
        if (source == null) {
            return 0;
        } else if (source.left == null && source.right == null) {
            return 1;
        } else {
            return 1 + Integer.max(this.dfs(source.left), this.dfs(source.right));
        }
    }
}

File: leetcode/python/0104. Maximum Depth of Binary Tree.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Time complexity: O(N), where N is the number of nodes
# Space complexity: O(N)

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.getNodes(root)
        
    def getNodes(self, root):
        if root == None:
            return 0
        elif root.left == None and root.right == None:
            return 1
        else:
            return 1 + max(self.getNodes(root.left), self.getNodes(root.right))